File: Design-Patterns-Using-Swift/Strategy/StrategyPatternSorter.swift
import Foundation

protocol Sorting {
    func sort(_ numbers: inout [Int])
}

class BubbleSort: Sorting {
    func sort(_ numbers: inout [Int]) {
        // Bubble Sort Algorithm
        print("Bubble Sort Algorithm executed.")
        let size = numbers.count
        for i in 0..<(size - 1) {
            for j in 0..<(size - i - 1) {
                if numbers[j] > numbers[j + 1] {
                    // Swapping
                    let temp = numbers[j]
                    numbers[j] = numbers[j + 1]
                    numbers[j + 1] = temp
                }
            }
        }
    }
}

class SelectionSort: Sorting {
    func sort(_ numbers: inout [Int]) {
        // Selection Sort Algorithm
        print("Selection Sort Algorithm executed.")
        let size = numbers.count
        for i in 0..<(size - 1) {
            var maxIndex = 0
            for j in 1..<(size - i) {
                if numbers[j] > numbers[maxIndex] {
                    maxIndex = j
                }
            }
            let temp = numbers[size - 1 - i]
            numbers[size - 1 - i] = numbers[maxIndex]
            numbers[maxIndex] = temp
        }
    }
}

class StrategyClass {
    private var sorter: Sorting

    init(_ algo: Sorting) {
        self.sorter = algo
    }

    func setSorter(_ algo: Sorting) {
        self.sorter = algo
    }

    func sort(_ a: inout [Int]) {
        self.sorter.sort(&a)
    }
}

// Client code.
var a = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10]
let s = StrategyClass(BubbleSort())
s.sort(&a)
print(a)

var b = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10]
s.setSorter(SelectionSort())
s.sort(&b)
print(b)

/*
Bubble Sort Algorithm executed.
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Selection Sort Algorithm executed.
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
*/

File: Design-Patterns-Using-CSharp/Strategy/StrategyPatternSorter.cs
using System;

interface ISorting
{
    void Sort(int[] numbers);
}

class BubbleSort : ISorting
{
    public void Sort(int[] numbers)
    {
        // Bubble Sort Algorithm
        Console.WriteLine("Bubble Sort Algorithm executed.");
        int size = numbers.Length;
        for (int i = 0; i < size - 1; i++)
        {
            for (int j = 0; j < size - i - 1; j++)
            {
                if (numbers[j] > numbers[j + 1])
                {
                    // Swapping
                    int temp = numbers[j];
                    numbers[j] = numbers[j + 1];
                    numbers[j + 1] = temp;
                }
            }
        }
    }
}

class SelectionSort : ISorting
{
    public void Sort(int[] numbers)
    {
        // Selection Sort Algorithm
        Console.WriteLine("Selection Sort Algorithm executed.");
        int size = numbers.Length;
        for (int i = 0; i < size - 1; i++)
        {
            int maxIndex = 0;
            for (int j = 1; j < size - i; j++)
            {
                if (numbers[j] > numbers[maxIndex])
                {
                    maxIndex = j;
                }
            }
            int temp = numbers[size - 1 - i];
            numbers[size - 1 - i] = numbers[maxIndex];
            numbers[maxIndex] = temp;
        }
    }
}

class StrategyClass
{
    private ISorting sorter;

    public StrategyClass(ISorting algo)
    {
        this.sorter = algo;
    }

    public void SetSorter(ISorting algo)
    {
        this.sorter = algo;
    }

    public void Sort(int[] a)
    {
        this.sorter.Sort(a);
    }
}

// Client code
public class StrategyPatternSorter
{
    public static void Main(string[] args)
    {
        int[] a = { 4, 5, 3, 2, 6, 7, 1, 8, 9, 10 };
        StrategyClass s = new StrategyClass(new BubbleSort());
        s.Sort(a);
        Console.WriteLine(string.Join(", ", a));

        int[] b = { 4, 5, 3, 2, 6, 7, 1, 8, 9, 10 };
        s.SetSorter(new SelectionSort());
        s.Sort(b);
        Console.WriteLine(string.Join(", ", b));
    }
}
